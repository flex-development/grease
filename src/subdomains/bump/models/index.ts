/**
 * @file Entry Point - Models
 * @module grease/bump/models
 */

export { default as RecommendedBump } from './recommended-bump.model'
