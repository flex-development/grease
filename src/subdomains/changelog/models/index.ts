/**
 * @file Entry Point - Models
 * @module grease/changelog/models
 */

export {
  default as AbstractChangelogAggregator
} from './changelog-aggregator.abstract'
export { default as ChangelogAggregator } from './changelog-aggregator.model'
export {
  default as ChangelogEntry,
  type ChangelogEntryDTO
} from './changelog-entry.model'
export {
  default as AbstractChangelogFormatter
} from './changelog-formatter.abstract'
export { default as ChangelogFormatter } from './changelog-formatter.model'
export { default as ChangelogInfile } from './changelog-infile.model'
export {
  default as ChangelogStream,
  type ChangelogStreamDTO
} from './changelog-stream.model'
export { default as CommitGroup } from './commit-group.model'
export { default as CommitType } from './commit-type.model'
