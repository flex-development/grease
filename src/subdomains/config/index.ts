/**
 * @file Entry Point - config
 * @module grease/config
 */

export { default as ConfigModule } from './config.module'
export type * from './interfaces'
export * from './models'
export * from './providers'
