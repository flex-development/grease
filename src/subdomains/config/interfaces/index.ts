/**
 * @file Entry Point - Interfaces
 * @module grease/config/interfaces
 */

export type { default as IGreaseConfig } from './grease-config.interface'
