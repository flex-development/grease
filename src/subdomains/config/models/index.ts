/**
 * @file Entry Point - Models
 * @module grease/config/models
 */

export { default as GreaseConfig } from './grease-config.model'
