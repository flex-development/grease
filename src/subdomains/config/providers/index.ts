/**
 * @file Entry Point - Providers
 * @module grease/config/providers
 */

export { default as ConfigService } from './config.service'
