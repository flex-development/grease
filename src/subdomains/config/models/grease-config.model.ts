/**
 * @file Models - GreaseConfig
 * @module grease/config/models/GreaseConfig
 */

import { ChangelogOperation } from '#src/changelog'
import type { IGreaseConfig } from '#src/config/interfaces'
import type { Commit } from '#src/git'
import { GlobalOptions } from '#src/options'
import { get, omit } from '@flex-development/tutils'
import { IsInstance, ValidateNested } from 'class-validator'

/**
 * Grease configuration options.
 *
 * @see {@linkcode GlobalOptions}
 * @see {@linkcode IGreaseConfig}
 *
 * @template T â€” Parsed commit type
 *
 * @class
 * @extends {GlobalOptions}
 * @implements {IGreaseConfig<T>}
 */
class GreaseConfig<T extends Commit = Commit> extends GlobalOptions
  implements IGreaseConfig<T> {
  /**
   * Changelog options.
   *
   * @see {@linkcode ChangelogOperation}
   *
   * @default
   *  new ChangelogOperation(opts.changelog)
   *
   * @public
   * @instance
   * @member {ChangelogOperation<T>} changelog
   */
  @IsInstance(ChangelogOperation)
  @ValidateNested()
  public changelog: ChangelogOperation<T>

  /**
   * Create a new options object.
   *
   * @see {@linkcode IGreaseConfig}
   *
   * @param {IGreaseConfig<T>?} [opts={}] - User options
   */
  constructor(opts: IGreaseConfig<T> = {}) {
    super(opts)

    this.changelog = new ChangelogOperation<T>({
      ...omit(get(opts, 'changelog'), ['from']),
      ...this
    })
  }
}

export default GreaseConfig
