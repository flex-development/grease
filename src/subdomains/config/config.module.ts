/**
 * @file ConfigModule
 * @module grease/config/ConfigModule
 */

import { LoggerService, ValidationService } from '#src/providers'
import { Module } from '@nestjs/common'
import { ConfigService } from './providers'

/**
 * Configuration module.
 *
 * @class
 */
@Module({
  exports: [ConfigService],
  providers: [ConfigService, LoggerService, ValidationService]
})
class ConfigModule {}

export default ConfigModule
