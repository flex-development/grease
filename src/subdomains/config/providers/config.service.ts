/**
 * @file Providers - ConfigService
 * @module grease/config/providers/ConfigService
 */

import pkg from '#pkg' assert { type: 'json' }
import type { IGreaseConfig } from '#src/config/interfaces'
import { GreaseConfig } from '#src/config/models'
import type { Commit } from '#src/git'
import { LoggerService } from '#src/log'
import { ValidationService } from '#src/providers'
import {
  ERR_MODULE_NOT_FOUND,
  ERR_UNKNOWN_FILE_EXTENSION,
  ERR_UNSUPPORTED_DIR_IMPORT,
  type NodeError
} from '@flex-development/errnode'
import type {
  Result as MkbuildResult,
  OutputExtension
} from '@flex-development/mkbuild'
import * as mlly from '@flex-development/mlly'
import * as pathe from '@flex-development/pathe'
import {
  DOT,
  cast,
  define,
  get,
  join,
  merge,
  omit,
  template,
  type EmptyObject,
  type Merge
} from '@flex-development/tutils'
import { Injectable } from '@nestjs/common'
import fg from 'fast-glob'
import json5 from 'json5'

/**
 * Grease configuration provider.
 *
 * @class
 */
@Injectable()
class ConfigService {
  /**
   * Configuration name.
   *
   * @public
   * @static
   * @readonly
   * @member {string} NAME
   */
  public static readonly NAME: string = pkg.name.replace(/.*\//, '')

  /**
   * Create a new configuration service.
   *
   * @see {@linkcode LoggerService}
   * @see {@linkcode ValidationService}
   *
   * @param {LoggerService} logger - Logger service
   * @param {ValidationService} validator - Validation service
   */
  constructor(
    protected readonly logger: LoggerService,
    protected readonly validator: ValidationService
  ) {
    this.logger = logger.withTag('config')
  }

  /**
   * Get the search pattern for config files.
   *
   * Supported file basenames:
   *
   * - `.grease`
   * - `.greaserc`
   * - `grease.config`
   *
   * Supported file extensions:
   *
   * - `.js`
   * - `.json`
   * - `.json5`
   * - `.jsonc`
   * - `.mjs`
   * - `.mts`
   * - `.ts`
   *
   * @protected
   * @static
   *
   * @return {string} Config file search pattern
   */
  protected static get SEARCH_PATTERN(): string {
    /**
     * Config file extensions.
     *
     * @const {string[]} exts
     */
    const exts: string[] = ['js', 'json', 'json5', 'jsonc', 'mjs', 'mts', 'ts']

    return template('{.{name},.{name}rc,{name}.config}.+({exts})', {
      exts: join(exts, '|'),
      name: ConfigService.NAME
    })
  }

  /**
   * Load a configuration file.
   *
   * @public
   * @async
   *
   * @template T - Parsed commit type
   *
   * @param {string} file - Path to config file
   * @param {IGreaseConfig<T>?} [opts={}] - Configuration overrides
   * @return {Promise<GreaseConfig<T>>} Validated configuration object
   * @throws {NodeError} If loading fails
   */
  public async load<T extends Commit = Commit>(
    file: string,
    opts: IGreaseConfig<T> = {}
  ): Promise<EmptyObject | GreaseConfig<T>> {
    this.logger.sync(opts)

    switch (true) {
      case opts.config === false:
        this.logger.verbose('load skipped')
        break
      case mlly.isFile(file):
        /**
         * URL of configuration file.
         *
         * @const {URL} url
         */
        const url: URL = mlly.toURL(file)

        /**
         * Config file extension.
         *
         * @const {string} ext
         */
        const ext: string = pathe.extname(url.href)

        /**
         * User configuration object.
         *
         * @var {IGreaseConfig<T>} config
         */
        let config: IGreaseConfig<T> = {}

        /**
         * Config file source content.
         *
         * @const {string} source
         */
        let source: string = <string>await mlly.getSource(url)

        // log load start
        this.logger.verbose(url.pathname)

        // parse config source based on extension
        switch (ext) {
          case '.json':
          case '.json5':
          case '.jsonc':
            config = json5.parse(source)
            break
          case '.js':
          case '.mjs':
            source = await mlly.resolveModules(source, { parent: url })
            config = get(await import(mlly.toDataURL(source)), 'default', {})
            break
          case '.mts':
          case '.ts':
            const { default: make } = await import('@flex-development/mkbuild')

            const [result] = <[MkbuildResult]>(await make({
              clean: false,
              configfile: false,
              cwd: pathe.dirname(file),
              dts: false,
              ext: cast<OutputExtension>(ext),
              format: 'esm',
              keepNames: true,
              pattern: pathe.basename(file),
              platform: 'node',
              source: DOT,
              write: false
            }))

            source = get(result, 'outputs.0.text')
            source = await mlly.resolveModules(source, { parent: url })
            config = get(await import(mlly.toDataURL(source)), 'default', {})
            break
          default:
            this.logger.fail('invalid extension', this.logger.colors.bold(ext))

            /**
             * Configuration loading error.
             *
             * @const {NodeError} error
             */
            const error: NodeError = new ERR_UNKNOWN_FILE_EXTENSION(ext, file)

            this.logger.trace(error)
            throw error
        }

        config = new GreaseConfig(this.merge(config, opts))
        this.logger.verbose(config)
        return this.validator.validate<GreaseConfig<T>>(cast(config))
      default:
        this.logger.fail('invalid config', this.logger.colors.bold(file))

        /**
         * Configuration loading error.
         *
         * @const {NodeError} error
         */
        const error: NodeError = mlly.isDirectory(file)
          ? new ERR_UNSUPPORTED_DIR_IMPORT(file, import.meta.url)
          : new ERR_MODULE_NOT_FOUND(file, import.meta.url, 'module')

        this.logger.trace(error)
        throw error
    }

    return {}
  }

  /**
   * Merge configuration objects.
   *
   * @see {@linkcode IGreaseConfig}
   *
   * @protected
   *
   * @template T - Configuration object type
   * @template U - Configuration overrides type
   *
   * @param {T} config - Configuration object
   * @param {U?} [opts={}] - Configuration overrides
   * @return {Merge<T, [U]>} Merged configuration object
   */
  protected merge<
    T extends IGreaseConfig = IGreaseConfig,
    U extends IGreaseConfig = IGreaseConfig
  >(config: T, opts: U = <U>{}): Merge<T, [U]> {
    config.changelog = merge({ ...config.changelog }, { ...opts.changelog })
    return merge<T, [U]>(config, <U>omit(opts, ['changelog']))
  }

  /**
   * Search for a configuration object.
   *
   * @see {@linkcode GreaseConfig}
   * @see {@linkcode IGreaseConfig}
   *
   * @public
   * @async
   *
   * @template T - Parsed commit type
   *
   * @param {IGreaseConfig<T>?} [opts={}] - Configuration overrides
   * @return {Promise<EmptyObject | GreaseConfig<T>>} Configuration options
   */
  public async search<T extends Commit = Commit>(
    opts: IGreaseConfig<T> = {}
  ): Promise<EmptyObject | GreaseConfig<T>> {
    const [file = ''] = await fg(ConfigService.SEARCH_PATTERN, opts)

    !file && define(opts, 'config', { value: false })
    return this.load(pathe.resolve(get(opts, 'cwd', ''), file), opts)
  }
}

export default ConfigService
