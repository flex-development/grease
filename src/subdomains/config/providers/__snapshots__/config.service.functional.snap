// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`functional:config/providers/ConfigService > #load > .js > should load configuration object 1`] = `
{
  "changelog": ChangelogOperation {
    "Aggregator": [Function ChangelogAggregator],
    "Commit": [Function Commit],
    "Formatter": [Function ChangelogFormatter],
    "actions": [],
    "colors": true,
    "config": true,
    "cwd": "$PWD/__fixtures__/pkg/major",
    "debug": true,
    "from": "",
    "infile": undefined,
    "issues": [],
    "outfile": undefined,
    "pr": [],
    "quiet": false,
    "releases": 1,
    "samefile": false,
    "tagprefix": "pkg-major@",
    "to": "HEAD",
    "types": [
      CommitType {
        "hidden": false,
        "section": ":sparkles: Features",
        "type": "feat",
      },
      CommitType {
        "hidden": false,
        "section": ":bug: Fixes",
        "type": "fix",
      },
      CommitType {
        "hidden": false,
        "section": ":mechanical_arm: Refactors",
        "type": "refactor",
      },
      CommitType {
        "hidden": false,
        "section": ":fire: Performance Improvements",
        "type": "perf",
      },
      CommitType {
        "hidden": false,
        "section": ":wastebasket: Reverts",
        "type": "revert",
      },
      CommitType {
        "hidden": false,
        "section": ":pencil: Documentation",
        "type": "docs",
      },
      CommitType {
        "hidden": false,
        "section": ":white_check_mark: Testing",
        "type": "test",
      },
      CommitType {
        "hidden": false,
        "section": ":package: Build",
        "type": "build",
      },
      CommitType {
        "hidden": false,
        "section": ":robot: Continuous Integration",
        "type": "ci",
      },
      CommitType {
        "hidden": false,
        "section": ":house_with_garden: Housekeeping",
        "type": "chore",
      },
      CommitType {
        "hidden": true,
        "section": "release",
        "type": "release",
      },
      CommitType {
        "hidden": true,
        "section": "style",
        "type": "style",
      },
      CommitType {
        "hidden": true,
        "section": "wip",
        "type": "wip",
      },
    ],
    "unstable": true,
    "write": false,
  },
  "colors": true,
  "config": true,
  "cwd": "$PWD/__fixtures__/pkg/major",
  "debug": true,
  "quiet": false,
  "tag": TagOperation {
    "colors": true,
    "config": true,
    "cwd": "$PWD/__fixtures__/pkg/major",
    "debug": true,
    "force": false,
    "message": "",
    "object": "HEAD",
    "push": false,
    "quiet": false,
    "remote": "origin",
    "sign": false,
    "tag": ".",
    "tagprefix": "pkg-major@",
    "unstable": true,
    "verify": false,
  },
  "tagprefix": "pkg-major@",
  "unstable": true,
}
`;

exports[`functional:config/providers/ConfigService > #load > .json > should load configuration object 1`] = `
{
  "changelog": ChangelogOperation {
    "Aggregator": [Function ChangelogAggregator],
    "Commit": [Function Commit],
    "Formatter": [Function ChangelogFormatter],
    "actions": [],
    "colors": true,
    "config": true,
    "cwd": "$PWD/__fixtures__/pkg/minor",
    "debug": true,
    "from": "",
    "infile": undefined,
    "issues": [],
    "outfile": undefined,
    "pr": [],
    "quiet": false,
    "releases": 1,
    "samefile": false,
    "tagprefix": "pkg-minor@",
    "to": "HEAD",
    "types": [
      CommitType {
        "hidden": false,
        "section": ":sparkles: Features",
        "type": "feat",
      },
      CommitType {
        "hidden": false,
        "section": ":bug: Fixes",
        "type": "fix",
      },
      CommitType {
        "hidden": false,
        "section": ":mechanical_arm: Refactors",
        "type": "refactor",
      },
      CommitType {
        "hidden": false,
        "section": ":fire: Performance Improvements",
        "type": "perf",
      },
      CommitType {
        "hidden": false,
        "section": ":wastebasket: Reverts",
        "type": "revert",
      },
      CommitType {
        "hidden": false,
        "section": ":pencil: Documentation",
        "type": "docs",
      },
      CommitType {
        "hidden": false,
        "section": ":white_check_mark: Testing",
        "type": "test",
      },
      CommitType {
        "hidden": false,
        "section": ":package: Build",
        "type": "build",
      },
      CommitType {
        "hidden": false,
        "section": ":robot: Continuous Integration",
        "type": "ci",
      },
      CommitType {
        "hidden": false,
        "section": ":house_with_garden: Housekeeping",
        "type": "chore",
      },
      CommitType {
        "hidden": true,
        "section": "release",
        "type": "release",
      },
      CommitType {
        "hidden": true,
        "section": "style",
        "type": "style",
      },
      CommitType {
        "hidden": true,
        "section": "wip",
        "type": "wip",
      },
    ],
    "unstable": true,
    "write": false,
  },
  "colors": true,
  "config": true,
  "cwd": "$PWD/__fixtures__/pkg/minor",
  "debug": true,
  "quiet": false,
  "tag": TagOperation {
    "colors": true,
    "config": true,
    "cwd": "$PWD/__fixtures__/pkg/minor",
    "debug": true,
    "force": false,
    "message": "",
    "object": "HEAD",
    "push": false,
    "quiet": false,
    "remote": "origin",
    "sign": false,
    "tag": ".",
    "tagprefix": "pkg-minor@",
    "unstable": true,
    "verify": false,
  },
  "tagprefix": "pkg-minor@",
  "unstable": true,
}
`;

exports[`functional:config/providers/ConfigService > #load > .json5 > should load configuration object 1`] = `
{
  "changelog": ChangelogOperation {
    "Aggregator": [Function ChangelogAggregator],
    "Commit": [Function Commit],
    "Formatter": [Function ChangelogFormatter],
    "actions": [],
    "colors": true,
    "config": true,
    "cwd": "$PWD/__fixtures__/pkg/patch",
    "debug": true,
    "from": "",
    "infile": undefined,
    "issues": [],
    "outfile": undefined,
    "pr": [],
    "quiet": false,
    "releases": 1,
    "samefile": false,
    "tagprefix": "pkg-patch@",
    "to": "HEAD",
    "types": [
      CommitType {
        "hidden": false,
        "section": ":sparkles: Features",
        "type": "feat",
      },
      CommitType {
        "hidden": false,
        "section": ":bug: Fixes",
        "type": "fix",
      },
      CommitType {
        "hidden": false,
        "section": ":mechanical_arm: Refactors",
        "type": "refactor",
      },
      CommitType {
        "hidden": false,
        "section": ":fire: Performance Improvements",
        "type": "perf",
      },
      CommitType {
        "hidden": false,
        "section": ":wastebasket: Reverts",
        "type": "revert",
      },
      CommitType {
        "hidden": false,
        "section": ":pencil: Documentation",
        "type": "docs",
      },
      CommitType {
        "hidden": false,
        "section": ":white_check_mark: Testing",
        "type": "test",
      },
      CommitType {
        "hidden": false,
        "section": ":package: Build",
        "type": "build",
      },
      CommitType {
        "hidden": false,
        "section": ":robot: Continuous Integration",
        "type": "ci",
      },
      CommitType {
        "hidden": false,
        "section": ":house_with_garden: Housekeeping",
        "type": "chore",
      },
      CommitType {
        "hidden": true,
        "section": "release",
        "type": "release",
      },
      CommitType {
        "hidden": true,
        "section": "style",
        "type": "style",
      },
      CommitType {
        "hidden": true,
        "section": "wip",
        "type": "wip",
      },
    ],
    "unstable": true,
    "write": false,
  },
  "colors": true,
  "config": true,
  "cwd": "$PWD/__fixtures__/pkg/patch",
  "debug": true,
  "quiet": false,
  "tag": TagOperation {
    "colors": true,
    "config": true,
    "cwd": "$PWD/__fixtures__/pkg/patch",
    "debug": true,
    "force": false,
    "message": "",
    "object": "HEAD",
    "push": false,
    "quiet": false,
    "remote": "origin",
    "sign": false,
    "tag": ".",
    "tagprefix": "pkg-patch@",
    "unstable": true,
    "verify": false,
  },
  "tagprefix": "pkg-patch@",
  "unstable": true,
}
`;

exports[`functional:config/providers/ConfigService > #load > .jsonc > should load configuration object 1`] = `
{
  "changelog": ChangelogOperation {
    "Aggregator": [Function ChangelogAggregator],
    "Commit": [Function Commit],
    "Formatter": [Function ChangelogFormatter],
    "actions": [],
    "colors": true,
    "config": true,
    "cwd": "$PWD/__fixtures__/pkg/premajor",
    "debug": true,
    "from": "",
    "infile": undefined,
    "issues": [],
    "outfile": undefined,
    "pr": [],
    "quiet": false,
    "releases": 1,
    "samefile": false,
    "tagprefix": "pkg-premajor@",
    "to": "HEAD",
    "types": [
      CommitType {
        "hidden": false,
        "section": ":sparkles: Features",
        "type": "feat",
      },
      CommitType {
        "hidden": false,
        "section": ":bug: Fixes",
        "type": "fix",
      },
      CommitType {
        "hidden": false,
        "section": ":mechanical_arm: Refactors",
        "type": "refactor",
      },
      CommitType {
        "hidden": false,
        "section": ":fire: Performance Improvements",
        "type": "perf",
      },
      CommitType {
        "hidden": false,
        "section": ":wastebasket: Reverts",
        "type": "revert",
      },
      CommitType {
        "hidden": false,
        "section": ":pencil: Documentation",
        "type": "docs",
      },
      CommitType {
        "hidden": false,
        "section": ":white_check_mark: Testing",
        "type": "test",
      },
      CommitType {
        "hidden": false,
        "section": ":package: Build",
        "type": "build",
      },
      CommitType {
        "hidden": false,
        "section": ":robot: Continuous Integration",
        "type": "ci",
      },
      CommitType {
        "hidden": false,
        "section": ":house_with_garden: Housekeeping",
        "type": "chore",
      },
      CommitType {
        "hidden": true,
        "section": "release",
        "type": "release",
      },
      CommitType {
        "hidden": true,
        "section": "style",
        "type": "style",
      },
      CommitType {
        "hidden": true,
        "section": "wip",
        "type": "wip",
      },
    ],
    "unstable": true,
    "write": false,
  },
  "colors": true,
  "config": true,
  "cwd": "$PWD/__fixtures__/pkg/premajor",
  "debug": true,
  "quiet": false,
  "tag": TagOperation {
    "colors": true,
    "config": true,
    "cwd": "$PWD/__fixtures__/pkg/premajor",
    "debug": true,
    "force": false,
    "message": "",
    "object": "HEAD",
    "push": false,
    "quiet": false,
    "remote": "origin",
    "sign": false,
    "tag": ".",
    "tagprefix": "pkg-premajor@",
    "unstable": true,
    "verify": false,
  },
  "tagprefix": "pkg-premajor@",
  "unstable": true,
}
`;

exports[`functional:config/providers/ConfigService > #load > .mjs > should load configuration object 1`] = `
{
  "changelog": ChangelogOperation {
    "Aggregator": [Function ChangelogAggregator],
    "Commit": [Function Commit],
    "Formatter": [Function ChangelogFormatter],
    "actions": [],
    "colors": true,
    "config": true,
    "cwd": "$PWD/__fixtures__/pkg/preminor",
    "debug": true,
    "from": "",
    "infile": undefined,
    "issues": [],
    "outfile": undefined,
    "pr": [],
    "quiet": false,
    "releases": 1,
    "samefile": false,
    "tagprefix": "pkg-preminor@",
    "to": "HEAD",
    "types": [
      CommitType {
        "hidden": false,
        "section": ":sparkles: Features",
        "type": "feat",
      },
      CommitType {
        "hidden": false,
        "section": ":bug: Fixes",
        "type": "fix",
      },
      CommitType {
        "hidden": false,
        "section": ":mechanical_arm: Refactors",
        "type": "refactor",
      },
      CommitType {
        "hidden": false,
        "section": ":fire: Performance Improvements",
        "type": "perf",
      },
      CommitType {
        "hidden": false,
        "section": ":wastebasket: Reverts",
        "type": "revert",
      },
      CommitType {
        "hidden": false,
        "section": ":pencil: Documentation",
        "type": "docs",
      },
      CommitType {
        "hidden": false,
        "section": ":white_check_mark: Testing",
        "type": "test",
      },
      CommitType {
        "hidden": false,
        "section": ":package: Build",
        "type": "build",
      },
      CommitType {
        "hidden": false,
        "section": ":robot: Continuous Integration",
        "type": "ci",
      },
      CommitType {
        "hidden": false,
        "section": ":house_with_garden: Housekeeping",
        "type": "chore",
      },
      CommitType {
        "hidden": true,
        "section": "release",
        "type": "release",
      },
      CommitType {
        "hidden": true,
        "section": "style",
        "type": "style",
      },
      CommitType {
        "hidden": true,
        "section": "wip",
        "type": "wip",
      },
    ],
    "unstable": true,
    "write": false,
  },
  "colors": true,
  "config": true,
  "cwd": "$PWD/__fixtures__/pkg/preminor",
  "debug": true,
  "quiet": false,
  "tag": TagOperation {
    "colors": true,
    "config": true,
    "cwd": "$PWD/__fixtures__/pkg/preminor",
    "debug": true,
    "force": false,
    "message": "",
    "object": "HEAD",
    "push": false,
    "quiet": false,
    "remote": "origin",
    "sign": false,
    "tag": ".",
    "tagprefix": "pkg-preminor@",
    "unstable": true,
    "verify": false,
  },
  "tagprefix": "pkg-preminor@",
  "unstable": true,
}
`;

exports[`functional:config/providers/ConfigService > #load > .mts > should load configuration object 1`] = `
{
  "changelog": ChangelogOperation {
    "Aggregator": [Function ChangelogAggregator],
    "Commit": [Function Commit],
    "Formatter": [Function ChangelogFormatter],
    "actions": [],
    "colors": true,
    "config": true,
    "cwd": "$PWD/__fixtures__/pkg/prepatch",
    "debug": true,
    "from": "",
    "infile": undefined,
    "issues": [],
    "outfile": undefined,
    "pr": [],
    "quiet": false,
    "releases": 1,
    "samefile": false,
    "tagprefix": "pkg-prepatch@",
    "to": "HEAD",
    "types": [
      CommitType {
        "hidden": false,
        "section": ":sparkles: Features",
        "type": "feat",
      },
      CommitType {
        "hidden": false,
        "section": ":bug: Fixes",
        "type": "fix",
      },
      CommitType {
        "hidden": false,
        "section": ":mechanical_arm: Refactors",
        "type": "refactor",
      },
      CommitType {
        "hidden": false,
        "section": ":fire: Performance Improvements",
        "type": "perf",
      },
      CommitType {
        "hidden": false,
        "section": ":wastebasket: Reverts",
        "type": "revert",
      },
      CommitType {
        "hidden": false,
        "section": ":pencil: Documentation",
        "type": "docs",
      },
      CommitType {
        "hidden": false,
        "section": ":white_check_mark: Testing",
        "type": "test",
      },
      CommitType {
        "hidden": false,
        "section": ":package: Build",
        "type": "build",
      },
      CommitType {
        "hidden": false,
        "section": ":robot: Continuous Integration",
        "type": "ci",
      },
      CommitType {
        "hidden": false,
        "section": ":house_with_garden: Housekeeping",
        "type": "chore",
      },
      CommitType {
        "hidden": true,
        "section": "release",
        "type": "release",
      },
      CommitType {
        "hidden": true,
        "section": "style",
        "type": "style",
      },
      CommitType {
        "hidden": true,
        "section": "wip",
        "type": "wip",
      },
    ],
    "unstable": true,
    "write": false,
  },
  "colors": true,
  "config": true,
  "cwd": "$PWD/__fixtures__/pkg/prepatch",
  "debug": true,
  "quiet": false,
  "tag": TagOperation {
    "colors": true,
    "config": true,
    "cwd": "$PWD/__fixtures__/pkg/prepatch",
    "debug": true,
    "force": false,
    "message": "",
    "object": "HEAD",
    "push": false,
    "quiet": false,
    "remote": "origin",
    "sign": false,
    "tag": ".",
    "tagprefix": "pkg-prepatch@",
    "unstable": true,
    "verify": false,
  },
  "tagprefix": "pkg-prepatch@",
  "unstable": true,
}
`;

exports[`functional:config/providers/ConfigService > #load > .ts > should load configuration object 1`] = `
{
  "changelog": ChangelogOperation {
    "Aggregator": [Function ChangelogAggregator],
    "Commit": [Function Commit],
    "Formatter": [Function ChangelogFormatter],
    "actions": [],
    "colors": true,
    "config": true,
    "cwd": "$PWD/__fixtures__/pkg/prerelease",
    "debug": true,
    "from": "",
    "infile": undefined,
    "issues": [],
    "outfile": undefined,
    "pr": [],
    "quiet": false,
    "releases": 1,
    "samefile": false,
    "tagprefix": "pkg-prerelease@",
    "to": "HEAD",
    "types": [
      CommitType {
        "hidden": false,
        "section": ":sparkles: Features",
        "type": "feat",
      },
      CommitType {
        "hidden": false,
        "section": ":bug: Fixes",
        "type": "fix",
      },
      CommitType {
        "hidden": false,
        "section": ":mechanical_arm: Refactors",
        "type": "refactor",
      },
      CommitType {
        "hidden": false,
        "section": ":fire: Performance Improvements",
        "type": "perf",
      },
      CommitType {
        "hidden": false,
        "section": ":wastebasket: Reverts",
        "type": "revert",
      },
      CommitType {
        "hidden": false,
        "section": ":pencil: Documentation",
        "type": "docs",
      },
      CommitType {
        "hidden": false,
        "section": ":white_check_mark: Testing",
        "type": "test",
      },
      CommitType {
        "hidden": false,
        "section": ":package: Build",
        "type": "build",
      },
      CommitType {
        "hidden": false,
        "section": ":robot: Continuous Integration",
        "type": "ci",
      },
      CommitType {
        "hidden": false,
        "section": ":house_with_garden: Housekeeping",
        "type": "chore",
      },
      CommitType {
        "hidden": true,
        "section": "release",
        "type": "release",
      },
      CommitType {
        "hidden": true,
        "section": "style",
        "type": "style",
      },
      CommitType {
        "hidden": true,
        "section": "wip",
        "type": "wip",
      },
    ],
    "unstable": true,
    "write": false,
  },
  "colors": true,
  "config": true,
  "cwd": "$PWD/__fixtures__/pkg/prerelease",
  "debug": true,
  "quiet": false,
  "tag": TagOperation {
    "colors": true,
    "config": true,
    "cwd": "$PWD/__fixtures__/pkg/prerelease",
    "debug": true,
    "force": false,
    "message": "",
    "object": "HEAD",
    "push": false,
    "quiet": false,
    "remote": "origin",
    "sign": false,
    "tag": ".",
    "tagprefix": "pkg-prerelease@",
    "unstable": true,
    "verify": false,
  },
  "tagprefix": "pkg-prerelease@",
  "unstable": true,
}
`;
